<h1>UK Accident Datasets</h1>
<h2>Analyst: Janelle Kaye Borabo</h2>
<hr>

In [42]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [43]:
uk_accdt= pd.read_csv('datasets\\uk_accident_data.csv')

In [87]:
uk_accdt

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [45]:
uk_accdt.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [46]:
uk_accdt.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [47]:
uk_accdt['Latitude']=uk_accdt['Latitude'].fillna(uk_accdt['Latitude'].mean())
uk_accdt['Longitude']=uk_accdt['Longitude'].fillna(uk_accdt['Longitude'].mean())
uk_accdt['Road_Surface_Conditions']=uk_accdt['Road_Surface_Conditions'].fillna(uk_accdt['Road_Surface_Conditions'].mode()[0])
uk_accdt['Road_Type']=uk_accdt['Road_Type'].fillna(uk_accdt['Road_Type'].mode()[0])
uk_accdt['Urban_or_Rural_Area']=uk_accdt['Urban_or_Rural_Area'].fillna(uk_accdt['Urban_or_Rural_Area'].mode()[0])
uk_accdt['Weather_Conditions']=uk_accdt['Weather_Conditions'].fillna(uk_accdt['Weather_Conditions'].mode()[0])

In [48]:
uk_accdt.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [49]:
uk_accdt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660679 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660679 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  660679 non-null  object 
 10  Road_Type                660679 non-null  object 
 11  Urban_or_Rural_Area      660679 non-null  object 
 12  Weather_Conditions       660679 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [50]:
uk_accdt['Index']=uk_accdt['Index'].astype('category')
uk_accdt['Accident_Severity']=uk_accdt['Accident_Severity'].astype('category')
uk_accdt['Accident Date']=uk_accdt['Accident Date'].astype('category')
uk_accdt['Light_Conditions']=uk_accdt['Light_Conditions'].astype('category')
uk_accdt['District Area']=uk_accdt['District Area'].astype('category')
uk_accdt['Road_Surface_Conditions']=uk_accdt['Road_Surface_Conditions'].astype('category')
uk_accdt['Road_Type']=uk_accdt['Road_Type'].astype('category')
uk_accdt['Urban_or_Rural_Area']=uk_accdt['Urban_or_Rural_Area'].astype('category')
uk_accdt['Weather_Conditions']=uk_accdt['Weather_Conditions'].astype('category')
uk_accdt['Vehicle_Type']=uk_accdt['Vehicle_Type'].astype('category')

In [51]:
uk_accdt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  category
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            660679 non-null  category
 3   Latitude                 660679 non-null  float64 
 4   Light_Conditions         660679 non-null  category
 5   District Area            660679 non-null  category
 6   Longitude                660679 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  category
 10  Road_Type                660679 non-null  category
 11  Urban_or_Rural_Area      660679 non-null  category
 12  Weather_Conditions       660679 non-null  category
 13  Vehicle_Type             660679 non-null  ca

<h1>Activity 2 UK Accident Questions</h1>
<hr>

<h1>1. Which type of vehicle has the most Serious accidents? </h1>

In [52]:
ser = uk_accdt[uk_accdt['Accident_Severity'] == 'Serious']

In [53]:
ser['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      66461
Van / Goods 3.5 tonnes mgw or under       4554
Motorcycle over 500cc                     3457
Bus or coach (17 or more pass seats)      3373
Goods 7.5 tonnes mgw and over             2321
Motorcycle 125cc and under                2031
Taxi/Private hire car                     1771
Motorcycle 50cc and under                 1014
Motorcycle over 125cc and up to 500cc     1014
Goods over 3.5t. and under 7.5t            857
Other vehicle                              767
Agricultural vehicle                       282
Minibus (8 - 16 passenger seats)           276
Pedal cycle                                 39
Data missing or out of range                 0
Ridden horse                                 0
Name: count, dtype: int64

<h2>Insight No.1</h2>
<h3>The car has the most number of serious accidents with 66461 accidents.</h3>

<h1>2. Which weather was the road surface has the most fatal accidents?</h1>
<hr>

In [54]:
fatal = uk_accdt[uk_accdt['Accident_Severity'] == 'Fatal']

In [55]:
weat = fatal.groupby(['Weather_Conditions','Road_Surface_Conditions']).size()

In [56]:
weat

Weather_Conditions     Road_Surface_Conditions
Fine + high winds      Dry                         103
                       Flood over 3cm. deep          1
                       Frost or ice                  3
                       Snow                          0
                       Wet or damp                  68
Fine no high winds     Dry                        5653
                       Flood over 3cm. deep          6
                       Frost or ice                125
                       Snow                         14
                       Wet or damp                1409
Fog or mist            Dry                          18
                       Flood over 3cm. deep          0
                       Frost or ice                  8
                       Snow                          0
                       Wet or damp                  56
Other                  Dry                          12
                       Flood over 3cm. deep          0
                  

<h2>Insight No.2</h2>
<h3>Fine no high winds, and Dry road surface has the most fatal accidents.</h3>

<h1>3. Is there a correlation between Number of Casualties and Number of Vehicles?</h1>
<hr>

In [57]:
uk_accdt['Number_of_Casualties'].corr(uk_accdt['Number_of_Vehicles'])

np.float64(0.2288888612692756)

<h2>Insight No.3</h2>
<h3>There is no correlation between Number of Casualties and Number of Vehicles.</h3>

<h1>4. Which year has the most accidents in the Urban Area?</h1>
<hr>|

In [58]:
urban = uk_accdt[uk_accdt['Urban_or_Rural_Area'] == 'Urban'] 

In [59]:
urban['Accident Date'].value_counts()

Accident Date
12/11/2021    456
13/07/2019    448
13/11/2021    447
06/12/2019    442
30/11/2019    442
             ... 
25/12/2021     97
25/12/2022     96
25/12/2019     94
10/01/2022     81
25/12/2020     68
Name: count, Length: 1461, dtype: int64

<h2>Insight no.4</h2>
<h3> The year 2021 had the most accidents in the Urban Area.</h3>

<h1>5. Which year has the most accidents in the Rural Area?</h1>
<hr>

In [60]:
rural = uk_accdt[uk_accdt['Urban_or_Rural_Area'] == 'Rural'] 

In [61]:
rural['Accident Date'].value_counts()

Accident Date
11/02/2021    333
14/08/2019    303
18/01/2019    277
31/01/2019    277
01/02/2019    271
             ... 
25/12/2019     63
25/12/2020     50
25/12/2022     49
30/12/2022     44
10/01/2022     42
Name: count, Length: 1461, dtype: int64

<h2>Insight no.5</h2>
<h3> The year 2021 had the most accidents in the Rural Area.</h3>

<h1>6. What is the number of Casualties	that have the fatal accident in Car type vechicle?</h1>
<hr>

In [62]:
casu = uk_accdt[(uk_accdt['Accident_Severity'] == "Fatal") & (uk_accdt['Vehicle_Type'] == 'Car')]

In [63]:
casu['Number_of_Casualties'].value_counts()

Number_of_Casualties
1     3667
2     1542
3      670
4      347
5      178
6       89
7       40
8       13
9        9
10       7
11       2
14       2
13       2
42       1
68       1
16       1
17       1
18       1
21       1
12       1
40       1
24       1
Name: count, dtype: int64

In [64]:
casu['Number_of_Casualties'].mean()

np.float64(1.8917439562110385)

<h2>Insight no.6</h2>
<h3> In the Car type vehicle, the number of casualties that have the most fatal accidents is 1 with 3367 fatal accidents.</h3>

<h1>7. Which type of vehicle has the most Slight accidents?</h1>
<hr>

In [65]:
slit = uk_accdt[uk_accdt['Accident_Severity'] == 'Slight']

In [66]:
slit['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      424954
Van / Goods 3.5 tonnes mgw or under       29139
Bus or coach (17 or more pass seats)      22180
Motorcycle over 500cc                     21861
Goods 7.5 tonnes mgw and over             14770
Motorcycle 125cc and under                13049
Taxi/Private hire car                     11368
Motorcycle over 125cc and up to 500cc      6537
Motorcycle 50cc and under                  6494
Goods over 3.5t. and under 7.5t            5172
Other vehicle                              4800
Minibus (8 - 16 passenger seats)           1671
Agricultural vehicle                       1644
Pedal cycle                                 152
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h2>Insight no.7</h2>
<h3> The car has the most number of Slight accidents with 424954 accidents.</h3>

<h1>8. Which type of vehicle has the most Fatal accidents?</h1>
<hr>

In [67]:
fatal = uk_accdt[uk_accdt['Accident_Severity'] == 'Fatal']

In [68]:
fatal['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      6577
Van / Goods 3.5 tonnes mgw or under       467
Motorcycle over 500cc                     339
Bus or coach (17 or more pass seats)      325
Goods 7.5 tonnes mgw and over             216
Motorcycle 125cc and under                189
Taxi/Private hire car                     155
Motorcycle over 125cc and up to 500cc     105
Motorcycle 50cc and under                  95
Other vehicle                              70
Goods over 3.5t. and under 7.5t            67
Minibus (8 - 16 passenger seats)           29
Agricultural vehicle                       21
Pedal cycle                                 6
Data missing or out of range                0
Ridden horse                                0
Name: count, dtype: int64

<h2>Insight no.8</h2>
<h3> The car has the most number of Fatal accidents with 6577 accidents.</h3>

<h1>9. Which District Area has the most Serious Accidents in Dry Road Surface Conditions?</h1>
<hr>

In [69]:
drt=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Serious') & (uk_accdt['Road_Surface_Conditions'] == 'Dry')]

In [70]:
drt['District Area'].value_counts()

District Area
Birmingham                   1053
Leeds                         839
Westminster                   761
Bradford                      531
Liverpool                     512
                             ... 
Western Isles                  14
Clackmannanshire               13
Orkney Islands                 12
Shetland Islands                8
London Airport (Heathrow)       7
Name: count, Length: 422, dtype: int64

<h2>Insight no.9</h2>
<h3> The District Area of Birmingham has the largest number of Serious accidents in Dry Road Surface Conditions, with 1053 accidents.</h3>

<h1>10. Which District Area has the most Fatal Accidents in Wet or Damp Road Surface Conditions?</h1>
<hr>

In [71]:
frt=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Fatal') & (uk_accdt['Road_Surface_Conditions'] == 'Wet or damp')]

In [72]:
frt['District Area'].value_counts()

District Area
Highland              40
Birmingham            35
Leeds                 32
Bradford              31
Aberdeenshire         27
                      ..
Clackmannanshire       0
Tamworth               0
Worthing               0
Berwick-upon-Tweed     0
Blyth Valley           0
Name: count, Length: 422, dtype: int64

<h2>Insight no.10</h2>
<h3> The District Area of Highland has the largest number of Fatal accidents in Wet or Damp Road Surface Conditions, with 40 accidents.</h3>

<h1>11. Which year has the most Slight Accidents in the District Area of Highland with Dry Road Surface Conditions?</h1>
<hr>

In [73]:
frtye=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Slight') & (uk_accdt['Road_Surface_Conditions'] == 'Dry') & (uk_accdt['District Area'] == 'Highland')]

In [74]:
frtye['Accident Date'].value_counts()

Accident Date
23/08/2019    5
04/06/2020    5
31/05/2021    4
30/07/2021    4
26/05/2020    4
             ..
12/08/2022    0
12/09/2022    0
12/02/2021    0
12/02/2022    0
01/01/2019    0
Name: count, Length: 1461, dtype: int64

<h2>Insight no.11</h2>
<h3> The year 2019 had the most slight accidents in the District Area of Highland with Dry Road Surface Conditions, with 5 accidents.</h3>

<h1>12. Which year has the most Serious Accidents in the District Area of Kensington and Chelsea with Frost or ice Road Surface Conditions?</h1>
<hr>

In [75]:
srtye=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Serious') & (uk_accdt['Road_Surface_Conditions'] == 'Frost or ice') & (uk_accdt['District Area'] == 'Kensington and Chelsea')]

In [76]:
srtye['Accident Date'].value_counts()

Accident Date
05/01/2021    1
01/01/2021    0
01/01/2022    0
01/02/2019    0
30/12/2019    0
             ..
31/03/2022    0
31/05/2019    0
31/05/2020    0
31/05/2021    0
01/01/2020    0
Name: count, Length: 1461, dtype: int64

<h2>Insight no.12</h2>
<h3> The year 2021 had the most serious accidents in the District Area of Kensington and Chelsea with Frost or ice Road Surface Conditions, with 1 accidents.</h3>

<h1>13. Which Light Conditions has the most Fatal Accidents using the Car type Vehicle?</h1>
<hr>

In [77]:
licar=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Fatal') & (uk_accdt['Vehicle_Type'] == 'Car')]

In [78]:
licar['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       3876
Darkness - lights lit          1410
Darkness - no lighting         1200
Darkness - lighting unknown      56
Darkness - lights unlit          35
Name: count, dtype: int64

<h2>Insight no.13</h2>
<h3> The Daylight had the most number of Fatal Accidents with 3876 accidents, using Car type Vehicle?.</h3>

<h1>14. Which Light Conditions has the most Fatal Accidents using the car-type Vehicle in Urban areas?</h1>
<hr>

In [81]:
liurcar=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Fatal') & (uk_accdt['Vehicle_Type'] == 'Car') & (uk_accdt['Urban_or_Rural_Area'] == 'Urban')]

In [84]:
liurcar['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       1284
Darkness - lights lit           946
Darkness - no lighting           28
Darkness - lighting unknown      20
Darkness - lights unlit          14
Name: count, dtype: int64

<h2>Insight no.14</h2>
<h3> The Daylight had the largest number of Fatal Accidents with 1284 accidents, using the Car type Vehicle in the Urban Area? </h3>

<h1>15. Which Number of Vehicles has the most Serious Accidents in Darkness light Conditions in the Birmingham District Area?</h1>
<hr>

In [88]:
dar=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Serious') & (uk_accdt['District Area'] == 'Birmingham') & (uk_accdt['Light_Conditions'] == 'Darkness - no lighting')]

In [90]:
dar['Number_of_Vehicles'].value_counts()

Number_of_Vehicles
1    4
2    2
Name: count, dtype: int64

<h2>Insight no.15</h2>
<h3> In the Number of Vehicles, the 1 had the largest number of Serious Accidents in Darkness light Conditions in the Birmingham District Area, with 4 accidents. </h3>

<h1>16. Which Number of Vehicles has the most Slight Accidents in Daylight Conditions in the Bradford District Area?</h1>
<hr>

In [93]:
darli=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Slight') & (uk_accdt['District Area'] == 'Bradford') & (uk_accdt['Light_Conditions'] == 'Daylight')]

In [94]:
darli['Number_of_Vehicles'].value_counts()

Number_of_Vehicles
2    2381
1    1063
3     341
4      71
5      18
6       3
7       2
8       2
Name: count, dtype: int64

<h2>Insight no.16</h2>
<h3> In the Number of Vehicles, the 2 had the largest number of Serious Accidents in Daylight Conditions in the Bradford District Area, with 2381 accidents. </h3>

<h1>17. Which Weather Conditions has the most Slight Accidents in Dry Road Surface Conditions, with only 1 Number of Casualties?</h1>
<hr>

In [95]:
waet=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Slight') & (uk_accdt['Number_of_Casualties'] == 1) & (uk_accdt['Road_Surface_Conditions'] == 'Dry')]

In [96]:
waet['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       292933
Fine + high winds          3150
Other                      1907
Raining no high winds       353
Fog or mist                 305
Raining + high winds         45
Snowing no high winds        36
Snowing + high winds          2
Name: count, dtype: int64

<h2>Insight no.17</h2>
<h3> The Fine no high winds had the most Slight Accidents in Dry Road Surface Conditions, with only 1 Number of Casualties and had 292933 accidents. </h3>

<h1>18. Which Weather Conditions has the most Fatal Accidents in Snowy Road Surface Conditions in the District Area of Highland?</h1>
<hr>

In [ ]:
hig=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Fatal') & (uk_accdt['Road_Surface_Conditions'] == 'Snow') & (uk_accdt['District Area'] == 'Highland')]

In [100]:
hig['Weather_Conditions'].value_counts()

Weather_Conditions
Snowing no high winds    1
Fine + high winds        0
Fog or mist              0
Fine no high winds       0
Other                    0
Raining + high winds     0
Raining no high winds    0
Snowing + high winds     0
Name: count, dtype: int64

<h2>Insight no.18</h2>
<h3> The Snowing no high winds, had the most Fatal Accidents in Snowy Road Surface Conditions, with only 1 accident in the District Area of Highland. </h3>

<h1>19. Which Road Type has the most Serious Accidents in Rural Area Conditions in Van Type Vehicle?</h1>
<hr>

In [101]:
van=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Serious') & (uk_accdt['Urban_or_Rural_Area'] == 'Rural') & (uk_accdt['Vehicle_Type'] == 'Van / Goods 3.5 tonnes mgw or under')]

In [102]:
van['Road_Type'].value_counts()

Road_Type
Single carriageway    1451
Dual carriageway       300
Roundabout              81
Slip road               10
One way street           4
Name: count, dtype: int64

<h2>Insight no.19</h2>
<h3> The Single carriageway had the most Serious Accidents in the Rural Area, with only 1451 accident using Van type vehicle. </h3>

<h1>20. Which Road Type has the most Fatal Accidents in Urban Area Conditions in Van Type Vehicle?</h1>
<hr>

In [104]:
vanni=  uk_accdt[(uk_accdt['Accident_Severity'] == 'Fatal') & (uk_accdt['Urban_or_Rural_Area'] == 'Urban') & (uk_accdt['Vehicle_Type'] == 'Van / Goods 3.5 tonnes mgw or under')]

In [105]:
vanni['Road_Type'].value_counts()

Road_Type
Single carriageway    114
Dual carriageway       36
Roundabout              6
One way street          3
Slip road               2
Name: count, dtype: int64

<h2>Insight no.20</h2>
<h3> The Single carriageway had the most Fatal Accidents in the Urban Area, with only 114 accident using Van type vehicle. </h3>